src/probability/kernel/disintegration.lean

```diff
@@ -0,0 +1,407 @@
+/-
+Copyright (c) 2023 Rémy Degenne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rémy Degenne
+-/
+import probability.kernel.cond_cdf
+import measure_theory.constructions.polish
+
+/-!
+# Disintegration of measures on product spaces
+
+Let `ρ` be a finite measure on `α × Ω`, where `Ω` is a standard Borel space. In mathlib terms, `Ω`
+verifies `[nonempty Ω] [topological_space Ω] [polish_space Ω] [measurable_space Ω] [borel_space Ω]`.
+
+For any measurable space `γ`, there exists a kernel `cond_kernel ρ : kernel α Ω` such that we
+have a disintegration of the constant kernel from `γ` with value `ρ`:
+`kernel.const γ ρ = (kernel.const γ ρ.fst) ⊗ₖ (kernel.prod_mk_left γ (cond_kernel ρ))`,
+where `ρ.fst` is the marginal measure of `ρ` on `α`. In particular,
+`ρ = ((kernel.const unit ρ.fst) ⊗ₖ (kernel.prod_mk_left unit (cond_kernel ρ))) ()`.
+
+In order to obtain a disintegration for any standard Borel space, we use that these spaces embed
+measurably into `ℝ`: it then suffices to define a suitable kernel for `Ω = ℝ`. In the real case,
+we define a conditional kernel by taking for each `a : α` the measure associated to the Stieltjes
+function `cond_cdf ρ a` (the conditional cumulative distribution function).
+
+## Main definitions
+
+* `probability_theory.cond_kernel ρ : kernel α Ω`: conditional kernel described above.
+
+## Main statements
+
+* `probability_theory.kernel.const_eq_comp_prod`:
+  `kernel.const γ ρ = (kernel.const γ ρ.fst) ⊗ₖ (kernel.prod_mk_left (cond_kernel ρ) γ)`
+* `probability_theory.measure_eq_comp_prod`:
+  `ρ = ((kernel.const unit ρ.fst) ⊗ₖ (kernel.prod_mk_left (cond_kernel ρ) unit)) ()`
+* `probability_theory.lintegral_cond_kernel`:
+  `∫⁻ a, ∫⁻ ω, f (a, ω) ∂(cond_kernel ρ a) ∂ρ.fst = ∫⁻ x, f x ∂ρ`
+
+## TODO
+
+* The finite measure hypothesis can be weakened to σ-finite. The proof uses the finite case.
```

Comment:
I don't think so. Think of Lebesgue measure on R x R. Its first marginal is too infinite. I think the right assumption is that rho.fst should be sigma-finite.

Reply:
Your example is very convincing, so I read my source again (foundations of modern probability, thm 3.4): the claim is that there is a disintegration for sigma-finite rho, but it would not use rho.fst in general, but another measure on the first space. If we want to use rho.fst then we need it to be sigma-finite indeed. Also in the general sigma-finite case the conditional kernel would not be Markov but only sigma-finite.
I removed the TODOs. It's better to have exact results later than to have misleading results now.